sklearn/neural_network/multilayer_perceptron.py

```diff
@@ -0,0 +1,1248 @@
+                             self.algorithm)
+
+    def _fit_lbfgs(self, X, y, activations, deltas, coef_grads,
+                   intercept_grads, layer_units):
+        # Store meta information for the parameters
+        self._coef_indptr = []
+        self._intercept_indptr = []
+        start = 0
+
+        # Save sizes and indices of coefficients for faster unpacking
+        for i in range(self.n_layers_ - 1):
+            n_fan_in, n_fan_out = layer_units[i], layer_units[i + 1]
+
+            end = start + (n_fan_in * n_fan_out)
+            self._coef_indptr.append((start, end, (n_fan_in, n_fan_out)))
+            start = end
+
+        # Save sizes and indices of intercepts for faster unpacking
+        for i in range(self.n_layers_ - 1):
+            end = start + layer_units[i + 1]
+            self._intercept_indptr.append((start, end))
+            start = end
+
+        # Run LBFGS
+        packed_coef_inter = _pack(self.coefs_,
+                                  self.intercepts_)
+
+        if self.verbose is True or self.verbose >= 1:
+            iprint = 1
+        else:
+            iprint = -1
+
+        optimal_parameters, self.loss_, d = fmin_l_bfgs_b(
+            x0=packed_coef_inter,
+            func=self._loss_grad_lbfgs,
+            maxfun=self.max_iter,
+            iprint=iprint,
+            pgtol=self.tol,
+            args=(X, y, activations, deltas, coef_grads, intercept_grads))
+
+        self._unpack(optimal_parameters)
+
+    def _fit_stochastic(self, X, y, activations, deltas, coef_grads,
+                        intercept_grads, layer_units, incremental):
+        rng = check_random_state(self.random_state)
+
+        if not incremental or not hasattr(self, '_optimizer'):
+            params = self.coefs_ + self.intercepts_
+            optimizer = stochastic_optimizers[self.algorithm]
+            self._optimizer = optimizer(params, self.learning_rate_init,
+                                        lr_schedule=self.learning_rate,
+                                        momentum=self.momentum,
+                                        nesterov=self.nesterovs_momentum,
+                                        power_t=self.power_t,
+                                        beta_1=self.beta_1, beta_2=self.beta_2,
+                                        epsilon=self.epsilon)
+
+        # early_stopping in partial_fit doesn't make sense
+        early_stopping = self.early_stopping and not incremental
+        if early_stopping:
+            X, X_val, y, y_val = train_test_split(
+                X, y, random_state=self.random_state,
+                test_size=self.validation_fraction)
+            if isinstance(self, ClassifierMixin):
+                y_val = self.label_binarizer_.inverse_transform(y_val)
+        else:
+            X_val = None
+            y_val = None
+
+        n_samples = X.shape[0]
+        batch_size = np.clip(self.batch_size, 1, n_samples)
```

Comment:
So for example, if I set self.batch_size to 10000 but n_samples is 100,
this will basically just do one batch. I think that deserves a warning at
least - and if the last minibatch is ragged it would warn as well. Maybe we
could handle this warning in the input validation instead, and avoid the
warning if the last minibatch is not the right size

On Wed, Oct 21, 2015 at 11:00 AM, Jiyuan Qian <notifications@github.com>
wrote:

> In sklearn/neural_network/multilayer_perceptron.py
> <https://github.com/scikit-learn/scikit-learn/pull/5214#discussion_r42634908>
> :
>
> > +                                        epsilon=self.epsilon)
> > +
> > +        # early_stopping in partial_fit doesn't make sense
> > +        early_stopping = self.early_stopping and not incremental
> > +        if early_stopping:
> > +            X, X_val, y, y_val = train_test_split(
> > +                X, y, random_state=self.random_state,
> > +                test_size=self.validation_fraction)
> > +            if isinstance(self, ClassifierMixin):
> > +                y_val = self.label_binarizer_.inverse_transform(y_val)
> > +        else:
> > +            X_val = None
> > +            y_val = None
> > +
> > +        n_samples = X.shape[0]
> > +        batch_size = np.clip(self.batch_size, 1, n_samples)
>
> It's not. The batch_size we take as argument in constructor is directly
> assigned to self.batch_size, but there can be cases when this value is
> larger than sample size, since batch_size is set when constructing the
> estimator and data is provided later when the user call fit or partial_fit
> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/5214/files#r42634908>.
>